flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -81,8 +85,283 @@ public void setPlannerTypeInferenceUtil(PlannerTypeInferenceUtil plannerTypeInfe
 		this.plannerTypeInferenceUtil = plannerTypeInferenceUtil;
 	}
 
+	/**
+	 * Registers a temporary system function.
+	 */
+	public void registerTemporarySystemFunction(
+			String name,
+			FunctionDefinition definition,
+			boolean ignoreIfExists) {
+		final String normalizedName = FunctionIdentifier.normalizeName(name);
+
+		if (definition instanceof UserDefinedFunction) {
+			try {
+				UserDefinedFunctionHelper.prepareInstance(config, (UserDefinedFunction) definition);
+			} catch (Throwable t) {
+				throw new ValidationException(
+					String.format(
+						"Could not register temporary system function '%s' due to implementation errors.",
+						name),
+					t);
+			}
+		}
+
+		if (!tempSystemFunctions.containsKey(normalizedName)) {
+			tempSystemFunctions.put(normalizedName, definition);
+		} else if (!ignoreIfExists) {
+			throw new ValidationException(
+				String.format(
+					"Could not register temporary system function. A function named '%s' does already exist.",
+					name));
+		}
+	}
+
+	/**
+	 * Drops a temporary system function. Returns true if a function was dropped.
+	 */
+	public boolean dropTemporarySystemFunction(
+			String name,
+			boolean ignoreIfNotExist) {
+		final String normalizedName = FunctionIdentifier.normalizeName(name);
+		final FunctionDefinition definition = tempSystemFunctions.remove(normalizedName);
+
+		if (definition == null && !ignoreIfNotExist) {
+			throw new ValidationException(
+				String.format(
+					"Could not drop temporary system function. A function named '%s' doesn't exist.",
+					name));
+		}
+
+		return definition != null;
+	}
+
+	/**
+	 * Registers a temporary catalog function.
+	 */
+	public void registerTemporaryCatalogFunction(
+			UnresolvedIdentifier unresolvedIdentifier,
```

Comment:
We're not being too consistent with the type of identifier that the `FunctionCatalog/CatalogManager` accepts (I'm guilty of that myself).

Some methods accept `UnresolvedIdentifier` e.g. `FunctionCatalog#registerTemporaryCatalogFunction`, `CatalogManager#dropTemporaryView`. Some resolved `ObjectIdentifier` e.g. `CatalogManager#createTemporaryTable`, `CatalogManager#createTable`.

I am not sure which one should we prefer. If we go with the `UnresolvedIdentifier` the benefit is that we always qualify in the `Catalog*`. The downside is that we would no longer qualify in the `*Operations`, (e.g. CreateTableOperation etc.), whereas we said that all Operations should be fully resolved...

We can create a follow up for that.

Reply:
I also observed these inconsistencies. I think the catalog manager can resolve identifiers internally and thus can just expose `UnresolvedIdentifier`. That's why all my newly added methods take the unresolved one.

We said that operation should be fully resolved but that was before the DDL. So things might have changed a bit. Maybe we should distinguish between resolved and unresolved operations. Let's discuss this in a follow up issue.

Reply:
Created https://issues.apache.org/jira/browse/FLINK-15859